---
title: "Background and Metadata"
teaching: 10
exercises: 0
questions:
- "What data are we using?"  
- "Why is this experiment important?"  
objectives:
- "Why study cuatro Cienegas?."
- "What is the difference between shotgun and amplicon metagenomics?"  
- "Understand the data set."
keypoints:
- "It’s important to record and understand your experiment’s metadata."  
- "Shotgun metagenomics sequences genomes of all organisms in a community."     
- "Amplicon metagenomics amplifies and secuences a selected region of DNA."   
---

# Background  
A metagenomic study is one in which we aim to understand the genetic composition 
of the species that coexist in a defined and preferably closed space. The microbiome
is the set of species that inhabit the same location. So, metagenomics studies try 
to gain knowledge about the microbiome composition of some place. Gut and nasopharinx 
human metagenomes are frequently used in biomedical research. Other examples of metagenome 
studies are environmental samples either terrestrial or marine. Finally, the agroindustry 
has interest in known trough metagenomics the microbiome that lives in plant roots and leaves.

<a href="{{ page.root }}/fig/metagenomic workflow.png">
  <img src="{{ page.root }}/fig/metagenomic workflow.png" alt="Cog Metagenome" />
</a>

## Cuatro Cienegas  
<a href="{{ page.root }}/fig/md-01-data-Stromatolites.jpeg">
  <img src="{{ page.root }}/fig/md-01-data-Stromatolites.jpeg" alt="Cuatro Cienegas" />
</a>

Cuatro Cienegas is an oasis in the mexican desert that can be a model for a 
[lost world](https://elifesciences.org/articles/38278).  Cuatro Cienegas shows
high content of sulphur and magnesium but little phosphorus and nutrients, conditions 
that resemble the ones found in the ancient seas. Some of the few live stromatolite
can be found at the lagoons of the Basin.  
  
## Shotgun and Amplicons    
A metagenomic study goes from the sampling design to the statistical analysis of the data sequenced. 
Mainly, two types of studies are carried out: amplicons and whole-genome sequencing (WGS) or shotgun. 
In amplicon studies, a region typical of a community of microorganisms is amplified and sequenced, 
for example the hypervariable regions of 16S in Bacteria, or ITS in Fungi.  In shotgun, random fragments 
of all genomes of all organisms in the sample are amplified and sequenced. Which is the most suitable? 
That depends on the question of the study.  

*Shotgun Metagenomics* Here we amplify and sequence random regions of the genome of all (or the vast majority (depends on their
abundance (and our sequencing depth))) of the organisms in the community.
It allows us to assemble regions of different genomes, with which we can investigate their genomic potential, i.e. what genes does it contain. It is more expensive because it requires greater depth of sequencing and requires more start gDNA because it does not include PCR amplification steps prior to library construction

*Amplicon 16s* To do metagenomics of amplicons, the first step is to amplify the region of interest through PCR reactions.
The primers we choose are VERY IMPORTANT. We always have biases that can change the results. Different primers show phylum specificity. 



> ## Amplicon or Shotgun? 
>
> Suppose you would like to compare the microbiome of sintomatic vs asintomatic pacients in certain disease.  
> Which type or metagenomics would you choose?  
> If you would to investigate variation of a certain gene in a microorganisms community would you use Amplicon
> or shotgun metagenomics?
> In which cases would you use Shotgun metagenomics?  
{: .discussion}





## Data 
  - The data we are going to use are part of an 
  [study about responses of a microbial community to nutrient enrichment](https://elifesciences.org/articles/49816)
  led by Jordan Okie in cuatro Cienegas. 
   
   - The study traces the genomic traits that madiated responses to nutrient enrichment
   in microbial communities in an oligotrophic, phosphorus-deficient pond
   from Cuatro Cienegas Basin (CCB), Mexico.
   
  - Results indicate that traits such as mean bacterial genome size, GC content, 
  total number of tRNA genes, total number of rRNA genes, and codon usage bias in 
  ribosomal protein sequences were all higher in the fertilized treatment.  
  
  - Metagenomic communities composition change in the control vs the 
  fertilized environment, where α proteobacteria seems to be more abundant.
  
  - Raw sequence data and metadata have been submitted to the NCBI Sequence Read 
  Archive, accessible through BioProject 
  [PRJEB22811](https://www.ncbi.nlm.nih.gov/sra/?term=PRJEB22811)
  
  - Other [Cuatro Cienegas datasets](https://www.mg-rast.org/mgmain.html?mgpage=search&search=cuatro%20cienegas) have 
  been generated by [Valeria souza](https://es.wikipedia.org/wiki/Valeria_Souza_Saldivar) group and can be found at MG-RAST database. 
  

> ## Metagenomic databases
> There are several public metagenomic
> [MG-RAST](https://www.mg-rast.org/index.html?stay=1)  
> [MGnify](https://www.ebi.ac.uk/metagenomics/), 
> [Marine Metagenomics Portal](https://mmp.sfb.uit.no/), 
> [Terrestrial Metagenome DB](https://webapp.ufz.de/tmdb/)  
> and the [GM Repo](https://gmrepo.humangut.info/home).   
{: .callout}


## Understanding the dataset  
Metadata is the data about data. This metadata describes information on one Cuatro Cienegas sample:

| Column           | Description                                |
|------------------|--------------------------------------------|
| file_name	          | JP4DASH2120627WATERAMPRESIZED.fasta				|
| investigation_type	       | metagenome		|
| seq_meth            | illumina	|
| project_description	        |  This project is a teaching project and uses data from Okie et al Elife 2020			|
| collection_date       | 2012-06-27 |
| country          | Mexico |
| feature         | pond water |
| latitude              | 26.8717055555556	|  
| longitude        | -102.14|  
| env_package  |	water|  
| depth	| 0.165 |   


> ## `.callout`
>
> An aside or other comment.
{: .callout}

> ## `.discussion`
>
> Discussion questions.
{: .discussion}



